### PR TITLE
[SPARK-9068][SQL] refactor the implicit type cast code

### PR DESCRIPTION
**Pull request by cloud-fan**
based on https://github.com/apache/spark/pull/7348

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -353,7 +353,7 @@ abstract class BinaryExpression extends Expression with trees.BinaryNode[Express
  * 2. Two inputs are expected to the be same type. If the two inputs have different types,
  *    the analyzer will find the tightest common type and do the proper type casting.
  */
-abstract class BinaryOperator extends BinaryExpression with ExpectsInputTypes {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: We have a rule in `HiveTypeCoercion` for `BinaryOperator`, extending `ExpectsInputTypes` only help us generate error message here. I think it's better to generate a different error message for `BinaryOperator`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 2]
```diff
@@ -713,27 +713,22 @@ object HiveTypeCoercion {
       @Nullable val ret: Expression = (inType, expectedType) match {
 
         // If the expected type is already a parent of the input type, no need to cast.
-        case _ if expectedType.isSameType(inType) => e
+        case _ if expectedType.acceptsType(inType) => e
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: Why we need `isSameType`? I think the logic should be:

1. first check whether the input type is acceptable(i.e. expectedType.acceptsType(inType) returns true). If it is, do nothing here.
2. If input type is not acceptable, follow special rules below.

It looks to me we only need `acceptsType`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 3]
```diff
@@ -107,13 +91,6 @@ private[sql] object TypeCollection {
     TimestampType, DateType,
     StringType, BinaryType)
 
-  /**
-   * Types that can be used in bitwise operations.
-   */
-  val Bitwise = TypeCollection(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: I checked with hive, hive's bitwise operations only accept integral type, boolean is not supported.

**Comment by cloud-fan** [comment]
cc @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 4]
```diff
@@ -369,17 +369,15 @@ abstract class BinaryOperator extends BinaryExpression with ExpectsInputTypes {
...
-          TypeCheckResult.TypeCheckSuccess
-        }
-      case TypeCheckResult.TypeCheckFailure(msg) => TypeCheckResult.TypeCheckFailure(msg)
+    // First check whether left and right have the same type, then check if the type is acceptable.
+    if (left.dataType != right.dataType) {
+      TypeCheckResult.TypeCheckFailure(s"differing types in '$prettyString' " +
+        s"(${left.dataType.simpleString} and ${right.dataType.simpleString}).")
+    } else if (!inputType.acceptsType(left.dataType)) {
+      TypeCheckResult.TypeCheckFailure(s"'$prettyString' accepts ${inputType.simpleString} type," +
+        s" not ${left.dataType.simpleString}")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: I think it's better to generate a different error message for `BinaryOperator`, as 2 children have same type and same expected type, we don't need to follow `ExpectsInputTypes` to report type error for both left and right.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 5]
```diff
@@ -691,7 +691,7 @@ object HiveTypeCoercion {
         // general implicit casting.
         val children: Seq[Expression] = e.children.zip(e.inputTypes).map { case (in, expected) =>
           if (in.dataType == NullType && !expected.acceptsType(NullType)) {
-            Cast(in, expected.defaultConcreteType)
+            Literal.create(null, expected.defaultConcreteType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: cast `NullType` to any type will result a literal null, so we can just write down the literal null here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala (hunk 2 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: there was a reason, but i can't remember why right now. let me think about it.

cc @marmbrus do you remember why?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 6]
```diff
@@ -747,15 +742,9 @@ object HiveTypeCoercion {
...
-        // use the current expression; otherwise, find the first one we can implicitly cast.
-        case (_, TypeCollection(types)) =>
-          if (types.exists(_.isSameType(inType))) {
-            e
-          } else {
-            types.flatMap(implicitCast(e, _)).headOption.orNull
-          }
+        // When we reach here, input type is not acceptable for any types in this type collection,
+        // try to find the first one we can implicitly cast.
+        case (_, TypeCollection(types)) => types.flatMap(implicitCast(e, _)).headOption.orNull
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: Is this correct? Imagine we have TypeCollection(LongType, StringType), and the input is StringType

Wouldn't this just cast input to a longtype?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: the first rule is: `case _ if expectedType.acceptsType(inType) => e`.
So when we reach here, input type is not acceptable for any types in this type collection. see my comments above.

**Comment by rxin** [comment]
LGTM - @marmbrus  and I can't really think of why anymore we needed it - maybe we don't need it anymore.  Going to merge this.

**Pull request closed without merging**